content/blog/2019-12-02-equivalence-modulo-inputs/index.md

```diff
@@ -0,0 +1,400 @@
+    for child in statement.children:
+      prune_visit(prog, child, coverage_set)
+
+def gen_variant(prog, coverage_set):
+  emi_variant = clone(prog)
+  for statement in emi_variant:
+    prune_visit(emi_variant, statement, coverage_set)
+
+  return emi_variant
+```
+
+`gen_variant` takes as input a seed program and the coverage set&mdash;the
+set of all statements executed for some input in the input set.
+It clones the program into `emi_variant` and then uses `prune_visit` to
+probabilistically delete unexecuted statements.
+
+With its EMI-variant generation algorithm outlined, we can now sketch
+the algorithm by which Orion validates C compilers.
+
+```Python
+def validate(compiler, prog, input_set):
+  # compile with no optimizations
+  out_prog = compiler.compile(prog, NO_OPTIMIZATION)
+
+  # generate reference output
+  in_out_set = [(i, out_prog.execute(i)) for i in input_set]
+
+  # get coverage info
+  # a statement is considered covered if it was executed
+  # by the program on any input
+  coverage_set = set()
+  for i in input_set:
+    coverage_set = union(coverage_set, coverage(prog, i))
+
+  for i in range(MAX_ITER):
+    emi_variant = gen_variant(prog, coverage_set)
+
+    for config in compiler.configurations:
+      out_emi_variant = compiler.compile(emi_variant, config)
+
+      # check if compiled EMI variant is equivalent over all inputs
+      for i, o in in_out_set:
+        # compiler is not EMI-valid!
+        if out_emi_variant.execute(i) != o:
+          report_bug(compiler, config, prog, emi_variant)
+```
+
+`validate` takes as input a compiler (`compiler`), a seed program (`prog`),
+and an input set (`input_set`).
+First, `validate` compiles `prog` with `compiler` using no optimizations,
+the output of which (`out_prog`) is then used to generate a reference set
+of outputs for `input_set`.
+Next, it uses a code coverage tool (`coverage`) to determine the set of
+executed statements over all inputs in `input_set`.
+In its "main loop," `validate` generates an EMI-variant of `prog` using
+the computed coverage information by calling `gen_variant`.
+For every relevant compiler configuration, it then compiles the EMI-variant
+and runs the output program over all inputs in `input_set` to check that it
+returns the same output as recorded in the reference set.
+If not, we flag the current compiler configuration as having a bug
+(`report_bug`).
+`validate` repeats this main loop for some number of iterations (`MAX_ITER`)
+to find more bugs using different EMI-variants.
+
+The authors note that the implementation effort for Orion is much less
+burdensome than other bug-finding tools such as Csmith:
+whereas Csmith is about 30K-40K lines of C++, Orion is only about 500 lines
+of shell scripts and 1K lines of C++.
+
+
+## Evaluation
+
+In order to evaluate EMI&mdash;even in its concrete implementation
+Orion&mdash;several questions must be answered:
+
+1. _What compilers (and compiler configurations) will be tested?_\
+\
```

Comment:
Irrelevantly, I think it's possible to get a new paragraph within a bulleted list in Markdown by indenting it to the same level as the first line of actual text (i.e., 3 spaces).